### PR TITLE
Regridding to a list of polygons (spatial averaging)

### LINKED ISSUE #11: Add Mesh support and a spatial averaging method
Averaging some gridded data over a polygon is something that often comes up in our work at Ouranosinc (and elsewhere for sure). As fair as I know, there is no easy way to do this with xarray datasets.  There is clisops's `subset_shape` (that was in `xclim.subset` before), but it creates a boolean mask with Trues when the grid centroid falls in the polygon.  What we want is a weights map, the weights being computed as the fraction of each gridcell that intersects the polygon. That is exactly what ESMF Regridding does with the conservative method. 

I suggest to read in a sequence of `shapely.geometry.Polygon` objects and generate a `ESMF.Mesh` from them.   Using these meshes with a few elements of many nodes, we can call regridding with a conservative method to get the weights. I have working code with ESMpy that does that and would be pleased to make PR if the idea is accepted.

Also, bonus of this issue, we could accept Geopandas Dataframe as input and output to the regridding object. That would add a lot of dependencies though. Shapely alone is more lightweight.

I'm guessing that my use case would require a new method. There are some issues with this:
- "MultiPolygon" objects would be splitted to their components. Weights could be summed afterwards to keep the hierarchy.
- Polygon with holes require a 2-pass treatment where weights computed from the holes are to be subtracted to the main "exterior-ring" weights.
Thus, I would suggest a `xesmf.intersection_weights()` method (or any other name).

### PR DESCRIPTION
**Pull request by aulemahal**
This PR adds support for a list of polygons as arguments of Regridder. When used in combination with a conservative method, the regridding output is equivalent to a spatial averaging of the input data on each polygon.

Uses `shapely` for the polygon implementation. The list of polygon is translated as a Mesh with multiple elements of various number of nodes. In the regridding process, the same behaviour than having `locstream` as output is used, thus regridded data has a `locations` dim that matches the order of the polygon list.

The `Regridder` will flatten the list of polygons (MultiPolygons object will be expanded) and will ignore potential holes in polygons. It makes its job easier, but is problematic for the end user. I thus created the `SpatialAverager` subclass that takes care of theses caveats. It divides the list of polygons into exteriors (remembering the "owners" of flattenned out MultiPolygons) and holes and then creates the weights using 2 Regridder calls. Weights are merged (flattenned polygons re-combined and holes subtracted) and normalized (requiring a conversion from COO sparse matrices to CSC). The SpatialAverager is then instantiated with these merged weights.

A new notebook was added and the "current limitations" page was update to highlight that this is not a efficient and elegant implementation of meshes, but a step forward.

Closes #11 

Previous comment:
<details>
STILL A DRAFT.
Before going forward, I wanted to discuss the following : Using GeoPandas we could make the use of this even more elegant. When ds_out is GeoDataFrame, the output could have the same index along "locations", instead of a simple integer one. Moreover, this would allow the implementation of a `spatial_averaging` method that could manage the holes and MutliPolygon cases.  Right now, when a MultiPolygon is encountered it is flattened out, split into its polygon parts. And holes are simply ignored. The downside would be all new dependencies (fiona and pyproj), it could be made optional tho?
</detais>

### CODE REVIEW

**Comment by aulemahal** [comment]
@andersy005  @huard @raphaeldussin  I'd like to have your input on adding `geopandas + fiona + pyproj` to the dependencies, either by default or as optional dependencies in order to have a cleaner PR. Accepting  `GeoDataframe`s would be more elegant than lists of shapely polygons, and would allow for a quite straightforward spatial averaging method : `xesmf.spatial_averages(ds, geo_df)` -> pd.Series with the same index as geo_df.

**Comment by raphaeldussin** [comment]
I'm all for using the libs that will make the code more elegant and thus easier to maintain!
Unsure about making the requirements optional or by default. Are we expecting potential conflicts between these
and the "core" dependencies?

I'm not super familiar with the Mesh part of ESMF, is the regridding possible from grid -> mesh and mesh -> grid ?
It looks like you can do mesh -> locstream, unsure about locstream -> mesh

I think we should try to allow all the possible in/out type combination that are supported by ESMF.

What do you think?

**Comment by huard** [comment]
The trade-off here is having to worry about these additional dependencies. In my experience, these are not "light" dependencies. I agree they would make the user-interface a lot friendlier. However, my concern is that they'll complicate the job of maintainers. 

Could we think about a compromise? For example, we could have a notebook that shows how to interface geopandas with xESMF. We run the notebook as part of the test suite, but geopandas is not part of the installation hard requirements.

**Comment by andersy005** [comment]
Another alternative is to make these additional packages soft dependencies by moving the `imports` of these additional packages inside the functions themselves

i.e.

```python
def func(ds):
  try:
    from shapely.geometry import MultiPolygon
  except ImportError as exc:
    message = f"shapely package is missing... Please install it via pip or conda: conda install shapely..."
    raise exc(message)
  # Do actual work here
  ...
```

rather than 

```python
from shapely.geometry import MultiPolygon
def func(ds):
  # Do actual work here
  ...
```

**Comment by aulemahal** [comment]
Thanks for the replies!
I think I will try to implement a "spatial_averaging" method without geopandas first, and make it optional if I really need it.

The last commits add the `polylist_in` option and we now have possibility to regrid grid or locstream to/from list of polygons, with the same method restrictions as for ESMF. (No restrictions for grid <-> Mesh, some for locstream <-> Mesh).

However, I was not able to make the nearests, bilinear and patch methods work with lists including polygons of more than 4 nodes. The documentation of ESMF says these should work, but I get a `rc=506` error:
`PET0 ESMCI_Array.C:1319 ESMCI::DD::setupSeqIndexFactorLo Invalid argument  - factorIndexList contains seqIndex outside Array bounds`
The very last commit uses a complex if statement to raise more meaningful error messages. May be the number of nodes is not the problem, but in my few tests that was the only difference that made the code break.
(P.S. : "nearest_s2d" works with polygons as input, "nearest_d2s" with polygons as output)

**Comment by aulemahal** [comment]
The `SpatialAverager` is ready for review! I edited the root comment to explain what this PR does.

I finally didn't need `geopandas` for the implmentation itself. However, I added it as a doc requirement and used it in my example notebook.

**Comment by aulemahal** [comment]
Caveats of this PR:

1. For changes in the Regridder obj. Some regridding methods involving Meshes that ESMF supports do not work here. The conditional statements in the init are quite complicated but here is the summary:

- With a list of polygons as input:
    * "bilinear" and "patch" never work, but should.
    * "nearest_d2s" only works when polygons have 4 or fewer nodes.
- With a list of polygons as output:
    * 'bilinrar', 'patch' and 'nearest_d2s' only work when polygons have 4 or fewer nodes.

In all cases, the error is as said above: `rc=506` error, `PET0 ESMCI_Array.C:1319 ESMCI::DD::setupSeqIndexFactorLo Invalid argument - factorIndexList contains seqIndex outside Array bounds`.

2. For the SpatialAverager. Spatial averaging should doable with a list of polygons as input. But the complexity of managing all the mutli-part geometries and holes cases was beyond the time I can put on this PR. I leave it for the future, if a need emerges.

**Review by huard — COMMENTED** [review]
First batch of comments.
I'll take some more time to try it out.

**Comment by huard** [comment]
My main concern is to see the `Regridder` class getting bloated by all these options for inputs and outputs. If later we implement support for meshes, the logic will become even messier. Also, I think we're abusing the intent for `ds_in` and `ds_out` by passing polygons instead of datasets as the name implies. 
Question: Is there a use case for `polygon_in` (that's not already covered by locstreams)? If not, we could decide to not support this and simplify the API. 

My suggestion (for discussion) would be to 
1. Create a `BaseRegridder` class only accepting ESMF objects (Grids, Mesh, LocStream).
2. Refactor `Regridder` to use `BaseRegridder`, including the current logic to convert datasets to either Grids or LocStream. 
3. Refactor `SpatialAverager` to use `BaseRegridder` and hold all the polygon handling logic. Note that the weights computation could probably overload `_compute_weights` instead of living in the `__init__`.

@mathause You might be interested to review this PR and provide your view on the implementation and the API.

**Review by mathause — COMMENTED** [review]
I am not familiar enough with ESMF to comment on the implementation but I would agree that the interface feels overloaded & using `polylist_in/out` seems unwieldy. @huard's suggestion seems reasonable to me.

- I find the name `locstream` extremely unintuitive - is this common knowledge for ESMF users?

- Note that the gridcells are not trusted in the notebook, therefore the HTML repr of xarray does not look nice, you may have to switch to the text repr (https://github.com/pangeo-data/xESMF/blob/c0b7d9ed9b102b26825e10ec7c85e0c02e521486/doc/notebooks/Spatial_Averaging.ipynb)

- The weights are exact and not an approximation, right? That is worth mentioning.

- How fast is the example? How long does it take when using all countries? (i.e. is it practicable in real life?) Do you know what the bottleneck is? (Creating the mesh or creating the weights?)

**Comment by raphaeldussin** [comment]
@mathause locstream is the name used in ESMF, for better or worse :)

**Comment by huard** [comment]
We should probably define the ESMF jargon in the docs though. Otherwise, there is a risk that users just don't know that a feature exists because it's called something else.

File: doc/limitations.rst
Comment (aulemahal) [reply doc/limitations.rst:25]: I was more trying to nuance the previous line by mentioning the polygon approach, but making it clear that this is not meant to be used for real meshes. If users have more elegant approaches, they should indeed open issues!

File: setup.py
Comment (aulemahal) [reply setup.py:19]: It could. I thought it was easy enough to install, but I don't know of all possible edge cases!

File: xesmf/frontend.py
Comment (aulemahal) [reply xesmf/frontend.py:139]: They are always ignored when transformed to a Mesh. The SpatialAverager takes care of creating two different Meshes, one for the exteriors and one for the holes. When using SpatialAverager, the poly list passed to `polys_to_ESMFmesh` does not contain any hole, so no warning is raised.

File: xesmf/frontend.py
Comment (aulemahal) [reply xesmf/frontend.py:171]: Oh yes, I should make that clearer.

File: xesmf/frontend.py
Comment (aulemahal) [reply xesmf/frontend.py:625]: Woups, indeed!

File: xesmf/frontend.py
Comment (aulemahal) [reply xesmf/frontend.py]: After testing it, there is indeed no errors. It seesm that ESMF allows meshes with overlapping elements! Removed it.

**Comment by aulemahal** [comment]
Here's a second version that answers comments of the reviewers. The main advice was to remove polygon handling from `Regridder` and move all ESM-only code to a `BaseRegridder` object, from which both `Regridder` and `SpatialAverager` can inherit. This was done, but demanded quite a bit more effort and refactoring then i originally thought... Hopefully, all changes are for the best for the future of xESMF...

Main changes, including many internal-breaking changes
-  `esmf_grid`, `esmf_locstream` and `esmf_mesh` were all transformed to OOP. xESMF now implements subclasses of ESMpy's `Grid`, `LocStream` and `Mesh` with the added class method `from_xarray` (`from_polygons` in `Mesh`).
- New class `BaseRegridder` that only accepts the three objects created above as `grid_in` and  `grid_out`, it does the weights computation and regridding, but not the conversion of `ds_in/out` to ESMF objects and the final xarray output formatting. 
`Regridder` takes care of that conversion and output (in `Regridder._format_xroutput`).
- This allows to have a `SpatialAverager` that inherits from `BaseRegridder` directly and also overrides `_format_xroutput` and `__repr__`.
- This means that parameters that were read from `ds_in/out` are now inferred from the `Grid/Mesh/LocStream` objects (shape and periodicity of longitude).
- To make it clearer that `LocStream` and `Mesh` input/output act the same (as 1D sequence of data points, instead of a 2D array), I renamed `BaseRegridder.locstream_out/in` to `BaseRegridder.sequence_in/out`. This can be easily reverted if its seen as unneeded. The input arguments of `Regridder` are still called `locstream_in/out`, so the breaking change is internal.

This means that Mesh regridding is indeed supported, but only if a ESMF Mesh is passed to `BaseRegridder`. The exposed `Regridder` does not have any option of that sort (`polylist_in/out` were removed).

@huard I wasn't able to find a clever way to move the weights managing code of `SpatialAverager` to `_compute_weights`. It still lives in the `__init__`...

@mathause With the removal of the useless overlapping check, things are a bit faster, but it is still slow. I tried the notebook example with all countries, simplified to a tolerance of 0.02°. I had to remove 8 of them as they were `invalid` according to shapely's rules and caused ESMF to fail (seems like a problem of my data source). Creating the SpatialAverager (and the weights) took 432 s, of which 424 s spend creating the two underlying Meshes (time of the 2 calls to `xesmf.backend.Mesh.from_polygons`).

**Comment by aulemahal** [comment]
Other change I forgot: in `esmf_regrid_finalize`, we destroy the regrid object, both fields and both grids. As I reuse the same grid twice in `SpatialAverager`, this was problematic. I commented the grid destroying lines. 
But it all this really needed? Quickly testing with memory profiler I found that:
```
del grid
```
has the same effect then
```
grid.destroy()
del grid
```
Not destroying the grid is not causing a memory leak.

**Comment by aulemahal** [comment]
Minor changes in the last commit, especially to remove all mentions of `esmf_grid`. 
There was one in the Backend notebook, I fixed it, but I can't run the notebook as I don't have ESMF build with MPI.  If somebody could run the notebook before we merge, that would be nice.

**Comment by huard** [comment]
Done. I've also made modifications to the spatial averaging noteboook:
- added explanations at the start
- removed references to North American grid
- added "country" coordinate instead of "polygon"

**Comment by huard** [comment]
@agstephens Do you want to take a look at this PR ? It implements grid averaging over polygons.

**Comment by agstephens** [comment]
> @agstephens Do you want to take a look at this PR ? It implements grid averaging over polygons.

@huard Thanks for pointing it out to me. I don't have much experience of ESMF or working with polygons in this way so I can't really provide useful comment. Tagging @sol1105 who might be interested from our project.

**Comment by sbiner** [comment]
I do not have anything to add/comment, I am not familiar with the modules and technique used.

This said I took a look at the [notebook](https://pangeo-xesmf--24.org.readthedocs.build/en/24/notebooks/Spatial_Averaging.html). Again a bit over my head but it sure looks impressive.  I also noticed what may be a typo: 

- In cell # 2 the comment talks about a 0.04 deg tolerance but the code shows a 0.02 value.

**Comment by mathause** [comment]
Shamless plug: you could also make a reference to [regionmask](https://github.com/mathause/regionmask) for computing regional averages over polygons with much better performance - if it is okay when the "weights" are an an approximation.

**Comment by huard** [comment]
@mathause I had thought about it, and agree it would be nice to make the link with regionmask. Could you please open a new issue and propose ideas to leverage the strengths of xesmf and regionmask?

**Comment by aulemahal** [comment]
I had a flash this afternoon and tried a modified version of the `Mesh.from_polygons` method. Using pre-allocated structured numpy arrays instead of dynamically growing lists, I was able to accelerate the mesh construction from polygons by a large factor. 

For the notebook example, time went from 1.7 s to 0.9s.
For the whole dataset (245 polygons, simplified to 0.02°, ~100 000 nodes), time went from 274 s to **35 s** !

However, I am using a numpy >=1.16  method in order to convert structured arrays to unstructured ones. I also tried a version with 2D arrays instead of structured ones and the speed up was nowhere near what I get here.

**Review by huard — APPROVED** [review]

**Comment by huard** [comment]
@bekozi This PR is inspired by what you've done in OCGIS to compute averages over polygons. Would you mind taking a look ?

**Comment by huard** [comment]
@raphaeldussin Review reminder ; )

There is another PR in the pipeline that started from this one. Merging this would make it easier to review.

File: xesmf/backend.py
Comment (raphaeldussin) [inline xesmf/backend.py:550]: what is the reason for commenting out these lines?
ESMF is prone to memory leaks if objects are not destroyed properly

File: xesmf/backend.py
Comment (raphaeldussin) [inline xesmf/backend.py:557]: same as above

File: xesmf/frontend.py
Comment (raphaeldussin) [inline xesmf/frontend.py:152]: no docstring for this option. Is it similar than the fix I introduced in PR #19 ?

**Review by raphaeldussin — CHANGES_REQUESTED** [review]

**Comment by raphaeldussin** [comment]
just a few minor things, also linting failed and so did readthedocs so I could not read the new notebook

File: xesmf/backend.py
Comment (aulemahal) [reply xesmf/backend.py:550]: With the new structure, we have the `BaseRegridder` object that takes ESMF objects as input. Thus, destroying them after the weight computation could cause issues for a user who created those themself.  
More specifically, I commented out these lines because `SpatialAverager` uses the same grid in two different `BaseRegridder` instances, so I instead of recreating it, I decided to reuse the same object.
A partial solution could be to destroy the grids where appropriate in `Regridder` and `SpatialAverager`.

File: xesmf/frontend.py
Comment (aulemahal) [reply xesmf/frontend.py:152]: Woups, I don't exactly remember, but this might be a leftover from my first `BaseRegridder` draft. I'll remove it. NaN's are indeed added to the weights the same way as done in #19, only the conditional statement has changed.

**Comment by aulemahal** [comment]
I can't understand the linting bug... I update pre-commit, did `autoupdate` and it finds nothing to do on my machine: the prenotebook hook passes. Moreover, what the pre-commit run suggests in the linting action seems ridiculous? I don't think there is a limit on notebook line lengths?

**Comment by aulemahal** [comment]
Voilà! 
As said elsewhere, the `prenotebook` hook needs `nodejs` to prettify the notebooks. It does nothing (silently) when `nodejs` isn't installed. Which it wasn't on mine (or @huard 's I presume), but is on the machines used for the checks.

File: xesmf/backend.py
Comment (raphaeldussin) [reply xesmf/backend.py:550]: ok let's see if this causes problems down the line. I will put an issue in to keep track of the potential problem

**Comment by raphaeldussin** [comment]
just a few more nitpicks: I would put the tutorial under intermediate, not beginners.
Id add to the imports:

```python
import warnings
warnings.filterwarnings("ignore")
```

and remove last empty cell (or should we remove keep-empty in the precommit yaml?)

**Review by raphaeldussin — APPROVED** [review]

**Comment by raphaeldussin** [comment]
I'm happy with the PR, great job @aulemahal 
@huard I let you press the green button, I got a turkey to go take care of ;)

**Pull request merged**
